proposals/0010-validating-resource-container-elements.md

```diff
@@ -0,0 +1,142 @@
+
+## Motivation
+
+Currently, in `clang\lib\CodeGen\CGHLSLRuntime.cpp`, under `calculateElementType` 
+there is a whitelist of valid RETs. 
+Anything that is not an int, uint, nor a floating-point type, will be rejected.
+The whitelist isn't broad enough, because it doesn't include the case where the RET 
+is user-defined. Ideally, a user should be able to determine how any user-defined 
+structure is invalid as an RET. Some system should be in place to more completely 
+enforce the rules for valid and invalid RETs, as well as provide useful information
+on why they are invalid.
+
+For example, `RWBuffer<bool> b : register(u4);` will not emit any error in DXC, 
+but will in clang-dxc, despite the fact that `bool` is a valid RET.
+
+## Proposed solution
+
+The proposed solution is to use some type_traits defined in the std library, create
+some custom type_traits that aren't defined there, and join them together to define a 
+set of conceptual constraints for any RET that is used. These conceptual constraints
+will be applied to every non-`RawBuffer` resource type that is defined, so that all
+non-`RawBuffer` HLSL resources have the same rules about which RETs are valid. 
+Validation will occur upon resource type instantiation. Additionally, certain 
+resource types are `RawBuffer` variants, such as `StructuredBuffer`. Such resource 
+types will have a `[[hlsl::raw_buffer]]` attribute in the attributed type. If this is
+detected, the rules for valid RETs will be loosened.
+
+## Detailed design
+
+In `clang\lib\Sema\HLSLExternalSemaSource.cpp`, `RWBuffer` is defined, along with 
+`RasterizerOrderedBuffer` and `StructuredBuffer`. It is at this point that the 
+`type_traits` should be incorporated into these resource declarations. All of the
+non-`RawBuffer` `type_traits` will be applied on each non-`RawBuffer` HLSL resource
+type. For every `type_trait` that is not true for the given RET, an associated error
+message will be emitted.
+
+The list of type_traits that define a valid non-`RawBuffer` RET are descsribed below:
+| type_trait | Description|
+|-|-|
+| `__is_complete_type` | An RET should either be a complete type, or a user defined type that has been completely defined. |
+| `__is_intangible_type` | An RET should not contain any handles with unknown sizes, i.e., should not be intangible. So, we should assert this type_trait is false. |
+| `__is_homogenous_aggregate` | RETs may be basic types, but if they are aggregate types, then all underlying basic types should be the same type. |
+| `__is_contained_in_four_groups_of_thirty_two_bits` | RETs should fit in four 32-bit quantities |
+
+Only `__is_complete_type` and `__is_intangible_type` are needed for `RawBuffer` RETs.
+
+* Examples:
+```
+struct x {
+	int i;
+};
+struct a {
+   int aa;
+   int ab;
+};
+struct b {
+   x bx;
+   int i;
+};
+struct c;
+struct d {
+  a ca;
+  float4 cb;
+};
+struct e {
+  int a;
+  int b;
+};
+struct f {
+  e x[2];
+  e y[2];
+};

```

Comment:
This would be a bit easier to follow if, instead of arbitrary struct names, they described what they contained. So `struct x` could be `struct oneInt` and `struct a` could be `struct twoInt` etc